packages/smithy-client/src/parse-utils.ts

```diff
@@ -222,6 +222,39 @@ export const expectString = (value: any): string | undefined => {
   throw new TypeError(`Expected string, got ${typeof value}`);
 };
 
+/**
+ * Asserts a value is a JSON-like object with only one non-null/non-undefined key and
+ * returns it.
+ *
+ * @param value A value that is expected to be an object with exactly one non-null,
+ *              non-undefined key.
+ * @return the value if it's a union, undefined if it's null/undefined, otherwise
+ *  an error is thrown.
+ */
+export const expectUnion = (value: unknown): { [key: string]: any } | undefined => {
+  if (value === null || value === undefined) {
+    return undefined;
+  }
+  const asObject = expectObject(value)!;
+
+  const setKeys = new Set<string>();
+  for (const [key, value] of Object.entries(asObject)) {
+    if (value !== null && value !== undefined) {
+      setKeys.add(key);
+    }
+  }
```

Comment:
Is the Set really needed? Since we are validating the object has exactly 1 key, and JS object doesn't allow multiple entries under same key, why not just `Object.keys(asObject)?.length !== 1`?

Reply:
You can have multiple keys where only one is not null - `{ a: 1, b: null, c: null }` is a valid union.

`Set` isn't being used here as a way to deduplicate. I just used it because order doesn't matter and object keys are expected to be unique anyway.

Reply:
Since `Object.entries` already returns a collection, it's looks unnecessary to create a Set explicitly. Will this do?

```TypeScript
const isDefined = (value: unknow): boolean => value !== null && value !== undefined;
if (Object.values(asObject).filter(isDefined).length !== 1) {
  throw new TypeError(`Unions must have exactly one non-null member`);
}
return asObject;
```

Reply:
I'd like to include which keys were non-null in the error message when more than one key is set, which I can't do if I just filter the values.

Reply:
The suggested way is more idiomatic. But we don't feel strong about it.